contributors/design-proposals/flexvolume-deployment.md

```diff
@@ -0,0 +1,161 @@
+
+
+## **Security Considerations**
+
+The Flexvolume driver directory can be continuously modified (accidentally or maliciously), making every` Probe()` call trigger a disk read, and `Probe()` calls could happen every couple of milliseconds and in bursts (i.e. lots of calls at first and then silence for some time). This may decrease kubelet's or controller manager's disk IO usage, impacting the performance of other system operations.
+
+As a safety measure, add a 1-second minimum delay between the processing of filesystem watch signals.
+
+
+## **Testing Plan**
+
+Add new unit tests in `plugin_tests.go` to cover new probing functionality and the heterogeneous plugin types in the plugins list.
+
+Add e2e tests that follow the user story. Write one for initial driver installation, one for an update for the same driver, one for adding another driver, and one for removing a driver.
+
+## **Recommended Driver Deployment Method**
+
+This section describes one possible method to automatically deploy Flexvolume drivers. The goal is that drivers must be deployed on nodes (and master when attach is required) without having to manually access any machine instance.
+
+Driver Installation:
+
+*   Alice is a storage plugin author and would like to deploy a Flexvolume driver on all node instances. She creates an image by copying her driver and the [deployment script](#driver-deployment-script) to a busybox base image, and makes her image available Bob, a cluster admin.
+*   Bob modifies the existing deployment DaemonSet spec with the name of the given image, and creates the DaemonSet.
+*   Charlie, an end user, creates volumes using the installed plugin.
+
+The user story for driver update is similar: Alice creates a new image with her new drivers, and Bob deploys it using the DaemonSet spec.
+
+### Driver Deployment Script
+
+This script assumes that only a *single driver file* is necessary, and is located at `/$DRIVER` on the deployment image.
+
+``` bash
+#!/bin/sh
+
+set -o errexit
+set -o pipefail
+
+VENDOR=k8s.io
+DRIVER=nfs
+
+driver_dir=$VENDOR${VENDOR:+"~"}${DRIVER}
+if [ ! -d "/flexmnt/$driver_dir" ]; then
+  mkdir "/flexmnt/$driver_dir"
+fi
+
+cp "/$DRIVER" "/flexmnt/$driver_dir/.$DRIVER"
+mv -f "/flexmnt/$driver_dir/.$DRIVER" "/flexmnt/$driver_dir/$DRIVER"
+
+while : ; do
+  sleep 3600
+done
+```
+
+### Deployment DaemonSet
+``` yaml
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+  name: flex-set
+spec:
+  template:
+    metadata:
+      name: flex-deploy
+      labels:
+        app: flex-deploy
+    spec:
+      containers:
+        - image: <deployment_image>
+          name: flex-deploy
+          securityContext:
+              privileged: true
+          volumeMounts:
+            - mountPath: /flexmnt
+              name: flexvolume-mount
+      volumes:
+        - name: flexvolume-mount
+          hostPath:
+            path: <host_driver_directory>
```

Comment:
It would be good if the author of this daemonset does not need to make assumptions about the host environment. For example, if `--volume-plugin-dir` varies for the nodes in the cluster it would still be good to author a single daemonset that install the flex volume correctly.

Could we make `--volume-plugin-dir` available in the downstream api so that it can be used as an env variable in the daemonset spec?